### PR TITLE
[2/3] build: integrate libsecp256k1

### LINKED ISSUE #2482: Port Bitcoin's key.cpp and pubkey.cpp code
While investigating https://github.com/bitcoin/bitcoin/pull/16878 I found that our `CKey` code differs significantly from upstream's, particularly in the `Set` and `Check` methods (we do not have these at all). This leads to upsteam's file `pubkey.cpp`, which we do not have either. It was created here: https://github.com/bitcoin/bitcoin/pull/5162. 

Many of these differences date to 2013 or 2014, which should be before our split from upstream, but could be vestigal Black/Peercoin code. For future porting and security, we should strive to bring our key code in line with Bitcoin.

### PR DESCRIPTION
**Pull request by div72**
Testing checklist:

- [x] Sync from zero.
- [x] Alerting.
- [x] Verifying/Signing messages.
- [x] Sending/Receiving TXs.
- [x] RPC commands.
- - [x] importprivkey
- - [x] importwallet
- - [x] dumpwallet
- - [x] getnewpubkey
- - [x] signmessage
- - [x] verifymessage
- - [x] addmultisigaddress
- - [x] makekeypair
- - [x] sendalert/sendalert2
- - [x] signrawtransaction

Fixes #2482.

### CODE REVIEW

File: src/gridcoin/contract/contract.cpp [hunk 1]
```diff
@@ -383,10 +383,7 @@ bool CheckLegacyContract(const Contract& contract, const CTransaction& tx)
 
     const uint256 body_hash = Hash(type_string, body.m_key, body.m_value);
 
-    CKey key;
```
Comment (jamescowens) [inline src/gridcoin/contract/contract.cpp:387]: Good. We have parameterized the master key...

File: src/gridcoin/contract/message.cpp [hunk 2]
```diff
@@ -26,7 +26,12 @@ namespace {
 //!
 bool SelectMasterInputOutput(CCoinControl& coin_control)
 {
-    const CTxDestination master_address = CWallet::MasterAddress().Get();
+    int best_block_height;
+    {
+        LOCK(cs_main);
```
Comment (jamescowens) [inline src/gridcoin/contract/message.cpp]: This is not necessary here. The only call scopes should be already protected by cs_main (for example SendVoteContract). You can add EXCLUSIVE_LOCKS_REQUIRED statements in the call chain to allow Clang to do automated checking if you want, which would be a really good idea. We need to put these throughout our code anyway. I got a good start on the scraper and related functions in an earlier PR.

**Review by jamescowens — COMMENTED** [review]
I know this is not done yet, but wanted to put up the few comments I had on what is there already.

File: src/gridcoin/contract/message.cpp (hunk 2 above)
Comment (div72) [reply src/gridcoin/contract/message.cpp]: I agree, but I think we should wait a bit and separate the cs_main before going through the codebase and applying thread safety attributes.

**Comment by div72** [comment]
@jamescowens b970cda1b4c4c7577c152cd6cf7251dd55c77906 merges the alert key and the master key in to the same key. This is fine on mainnet since the keys are same but the alert key and master key for the testnet is different. There are three possible solutions I can think of:

1. Re-split the alert and master keys. I am not in favour of this as I think it makes more sense for the keys to be one.
2. Add a block height conditional entry to change testnet master key to the alert key. The private key for this key is public which means others will also be able use it.
3. Change the testnet master key to the mainnet master key. This will prevent others from using the alert functionality but is the most simple option to implement as the non-alert key was the same key.

What do you think?

**Comment by jamescowens** [comment]
I have a different point of view on this. I think the keys should be separate. They serve different purposes.

**Comment by div72** [comment]
In that case, shouldn't mainnet's keys be separated too?

**Comment by jamescowens** [comment]
Yes. We need to change all of the keys out anyway.

**Comment by jamescowens** [comment]
That should not be done in this PR though. We should simply put the structure to have them separate for right now in both mainnet and testnet.

**Comment by jamescowens** [comment]
Upon request, compiled at https://github.com/gridcoin-community/Gridcoin-Research/pull/2492/commits/99fbaff12d41f40070148dba0d50c44f4d5f6c76 and brought up testnet node with researcher CPID.
1. Startup was unremarkable
2. Wallet unlocked successfully
3. Normal send transaction was successful
4. MRC request send and payment successful
5. Successfully staked CBR+RR

File: src/chainparams.h [hunk 3]
```diff
@@ -46,6 +46,14 @@ class CChainParams
     const Consensus::Params& GetConsensus() const { return consensus; }
     const CMessageHeader::MessageStartChars& MessageStart() const { return pchMessageStart; }
     const std::vector<unsigned char>& AlertKey() const { return vAlertPubKey; }
+    const std::vector<unsigned char>& MasterKey(int height) const {
+        for (auto it = masterkeys.rbegin(); it != masterkeys.rend(); ++it) {
+            if (it->first <= height) return it->second;
+        }
+
+        assert(false && "No master key specified or height is negative.");
+        std::abort();
```
Comment (jamescowens) [inline src/chainparams.h]: assert actually calls __assert_fail, which prints the assert message and aborts, so I think the std::abort() is a duplicate.

**Review by jamescowens — APPROVED** [review]
tACK (subject to the one trivial comment). Really nice job.

File: src/chainparams.h (hunk 3 above)
Comment (div72) [reply src/chainparams.h]: Fixed.

**Pull request merged**